### PR TITLE
HIP-410 - Ethereum Transaction Interop

### PR DESCRIPTION
**Pull request by shemnon**
**Description**:

Implemente [HIP-410](https://hips.hedera.com/hip/hip-410)

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthTxData.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+		byte[] s;
+		if (rlpItem.isList()) {
+			// frontier TX
+			type = EthTransactionType.LEGACY_ETHEREUM;
+			List<RLPItem> rlpList = rlpItem.asRLPList().elements();
+			if (rlpList.size() != 9) {
+				return null;
+			}
+			nonce = rlpList.get(0).asLong();
+			gasPrice = rlpList.get(1).asBytes();
```
Comment (steven-sheehy) [inline hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthTxData.java:69]: Per [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559), should maxPriorityGas and maxGas be set to gasPrice for legacy transactions when they normalize the transaction?

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	@Override
+	public ResponseCodeEnum validateSemantics(TxnAccessor accessor) {
+		var ethTxData = spanMapAccessor.getEthTxDataMeta(accessor);
+		var txBody = getOrCreateTransactionBody(accessor);
+
+		if (ethTxData.chainId().length == 0 || Arrays.compare(chainId, ethTxData.chainId()) != 0) {
+			return ResponseCodeEnum.WRONG_CHAIN_ID;
+		}
```
Comment (dimitar-dinev) [inline hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:142]: Is the span map populated with the data at the time of the semantic prechecks?

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+		if (accessor.getExpandedSigStatus() == ResponseCodeEnum.OK) {
+			// this is not precheck, so do more involved checks
+			maybeUpdateCallData(accessor, ethTxData, accessor.getTxn().getEthereumTransaction());
+			var ethTxSigs = getOrCreateEthSigs(txnCtx.accessor(), ethTxData);
+			var callingAccount = aliasManager.lookupIdBy(ByteString.copyFrom(ethTxSigs.address()));
+			if (callingAccount == null) {
+				return ResponseCodeEnum.INVALID_ACCOUNT_ID; 
+			}
```
Comment (dimitar-dinev) [inline hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:151]: Looking at the [return value of `aliasManager.lookupIdBy()`](https://github.com/hashgraph/hedera-services/blob/09d41f87924346d09653e6ef9cda10e468e13b2d/hedera-node/src/main/java/com/hedera/services/ledger/accounts/AliasManager.java#L166-L168), should we check against `MISSING_NUM` and not `null`?

Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:166]: Fixed.

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -2,7 +2,7 @@ name: E2E
 
 on:
   pull_request:
-    branches: [ master, release/** ]
+    branches: [ master, release/**, eth-tx-interop ]
```
Comment (Nana-EC) [inline .github/workflows/e2e.yml:5]: Unsure, but I don't think you want to add this branch for e2e logic in the main branch github workflow right?

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+					.setFunctionParameters(ByteString.copyFrom(ethTxData.callData()))
+					.setGas(ethTxData.gasLimit())
+					.setAmount(ethTxData.value().divide(WEIBARS_TO_TINYBARS).longValueExact())
+					.setContractID(EntityIdUtils.contractIdFromEvmAddress(ethTxData.to())).build();
+			return TransactionBody.newBuilder().setContractCall(synthOp).build();
+		} else {
+			final var autoRenewPeriod =
+					Duration.newBuilder().setSeconds(globalDynamicProperties.minAutoRenewDuration()).build();
+			var synthOp = ContractCreateTransactionBody.newBuilder()
+					.setInitcode(ByteString.copyFrom(ethTxData.callData()))
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:240]: Q: Is this prehydration? If so then it'd always be empty when the initCode is in a file.
Should there be a conditional to populate fileID on empty callData?

Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:240]: * This synthetic operation is not what is persisted in the record file.  It's only used to drive the transition logic.
* The ethTxData is the rehydrated version, if it had a fileId it was inserted
* We require a rehydrated EthTxData for signature validation to work.

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+		var callingAccount = aliasManager.lookupIdBy(ByteString.copyFrom(ethTxSigs.address()));
+
+		if (syntheticTxBody.hasContractCall()) {
+			contractCallTransitionLogic.doStateTransitionOperation(syntheticTxBody, callingAccount.toId(), true);
+		} else if (syntheticTxBody.hasContractCreateInstance()) {
+			final var synthOp =
+					addInheritablePropertiesToContractCreate(syntheticTxBody.getContractCreateInstance(), callingAccount.toGrpcAccountId());
+			syntheticTxBody = TransactionBody.newBuilder().setContractCreateInstance(synthOp).build();
+			spanMapAccessor.setEthTxBodyMeta(txnCtx.accessor(), syntheticTxBody);
+			contractCreateTransitionLogic.doStateTransitionOperation(syntheticTxBody, callingAccount.toId(), true);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java]: The patch to add the syntheitc op (#3289) fixed this.

File: hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java:250]: Gotcha.

**Review by tinker-michaelj — COMMENTED** [review]
LGTM

**Review by tinker-michaelj — APPROVED** [review]
🥂

**Review by tinker-michaelj — APPROVED** [review]
😌

**Pull request merged**